### ISSUE TITLE
Multiple ArgoCD Instances in different Namespaces

### ISSUE THREAD

**Issue by sylus** [argoproj-labs/argocd-operator#337]
Using the watched_namespace environment variable I set ArgoCD to watch multiple namespaces such as "argocd-operator-system,org-fdi-system". However while ArgoCD does detect all of the ArgoCD instances I get an issue with the following message:

```sh
unable to get: %v because of unknown namespace for the cache
```

It seems this is related to the following upstream bug:

- https://github.com/kubernetes-sigs/controller-runtime/issues/934

I was just curious as to if any other people have came across this issue and maybe if I myself am doing something wrong. Appreciate any insight you can provide!

**Comment by iam-veeramalla** [comment]
Hi @sylus , Thanks for raising issue with us. 

v0.0.15 natively supports this without need to add any watched_namespace environment variable or any other settings. However, I would recommend you to look at the list of issues users faced while upgrading.
https://github.com/argoproj-labs/argocd-operator/issues?q=is%3Aissue+is%3Aopen+label%3Aupgrade-issues

**Comment by sylus** [comment]
Hey @iam-veeramalla thanks for the response I think I am on 

quay.io/argoprojlabs/argocd-operator@sha256:38254c8d53172993cb51c0b4c72645757ba36e7e8e3b1251683b00cede47fa7c

Which to my knowledge is that v0.0.15, I believe i first tried to not set the watch namespace, but maybe I missed something I will try again today and report back!

**Comment by sylus** [comment]
Hey @iam-veeramalla so I must be doing something wrong but not sure. 

I manually made another ArgoCD instance and the operator detects it but then it just stalls.

![image](https://user-images.githubusercontent.com/984304/121949030-04540a80-cd26-11eb-88e8-67fbf00fd79b.png)

```
{"level":"info","ts":1623698842.32567,"logger":"controller_argocd","msg":"Reconciling ArgoCD","namespace":"argocd-operator-system","name":"fdi"}
{"level":"info","ts":1623698842.3257105,"logger":"controller_argocd","msg":"reconciling status"}
{"level":"info","ts":1623698842.3258228,"logger":"controller_argocd","msg":"reconciling roles"}
{"level":"info","ts":1623698848.7115517,"logger":"controller_argocd","msg":"reconciling rolebindings"}
{"level":"info","ts":1623698857.88098,"logger":"controller_argocd","msg":"reconciling service accounts"}
{"level":"info","ts":1623698864.16943,"logger":"controller_argocd","msg":"reconciling certificate authority"}
{"level":"info","ts":1623698864.1694777,"logger":"controller_argocd","msg":"reconciling CA secret"}
{"level":"info","ts":1623698864.1695187,"logger":"controller_argocd","msg":"reconciling CA config map"}
{"level":"info","ts":1623698864.16954,"logger":"controller_argocd","msg":"reconciling secrets"}
{"level":"info","ts":1623698864.2561843,"logger":"controller_argocd","msg":"reconciling config maps"}
{"level":"info","ts":1623698865.3836024,"logger":"controller_argocd","msg":"reconciling services"}
{"level":"info","ts":1623698865.3837276,"logger":"controller_argocd","msg":"reconciling deployments"}
{"level":"info","ts":1623698867.8806946,"logger":"controller_argocd","msg":"reconciling statefulsets"}
{"level":"info","ts":1623698867.8807452,"logger":"controller_argocd","msg":"existing argocd export not found, skipping import"}
{"level":"info","ts":1623698868.5787032,"logger":"controller_argocd","msg":"reconciling autoscalers"}
{"level":"info","ts":1623698868.6800144,"logger":"controller_argocd","msg":"reconciling ingresses"}
{"level":"info","ts":1623698868.6801383,"logger":"controller_argocd","msg":"reconciling prometheus"}
{"level":"info","ts":1623698868.680242,"logger":"controller_argocd","msg":"Reconciling ArgoCD","namespace":"argocd-operator-system","name":"org-fdi-system"}
```

Now it has said nothing for the past 10 minutes.

**Comment by sylus** [comment]
Hmm I think I noticed something for some reason if I only have the "FDI" ArgoCD Instance and no other spec I still get this:

```
{"level":"debug","ts":1623702098.6278584,"logger":"controller-runtime.controller","msg":"Successfully Reconciled","controller":"argocd-controller","request":"argocd-operator-system/fdi"}
{"level":"info","ts":1623702098.6279306,"logger":"controller_argocd","msg":"Reconciling ArgoCD","namespace":"argocd-operator-system","name":"org-fdi-system"}
{"level":"debug","ts":1623702098.6279554,"logger":"controller-runtime.controller","msg":"Successfully Reconciled","controller":"argocd-controller","request":"argocd-operator-system/org-fdi-system"}
```

I don't know where it is still getting Reconciling ArgoCD","namespace":"argocd-operator-system","name":"org-fdi-system" since it doesn't exist.

**Comment by sylus** [comment]
Please let me know if any more information i can provide @iam-veeramalla  :D

**Comment by iam-veeramalla** [comment]
> Please let me know if any more information i can provide @iam-veeramalla :D

You dont have to modify anything in the installation.
1. Install the argocd-operator in the `openshift-operators` ns.
2. Created ArgoCD 1 and ArgoCD 2 in different namespaces.

**Comment by sylus** [comment]
@iam-veeramalla I am not using openshift so I put it in the argocd-operator-system NS, beyond I made no adjustments.

**Comment by iam-veeramalla** [comment]
> @iam-veeramalla I am not using openshift so I put it in the argocd-operator-system NS, beyond I made no adjustments.
> 
> My thinking is this multi namespace logic was added only for OpenShift looking at the commit history?
> 
> Also greatly appreciate you taking the time to help me :D

ah, sorry I missed that. Let me try to get a k8s cluster and see whats going there :)

**Comment by sylus** [comment]
@iam-veeramalla I was trying to update myself to fix the cache bug which I think is needed but updating controller-manager also updates a lot of other deps and this doesn't include the operator-framework updates as well.

Seems like a fair bit of work to get this fixed.

**Comment by sylus** [comment]
@iam-veeramalla ok it wasn't so bad I just had to update to latest controller-runtime and then was able to specify multiple namespaces and everything worked :D

**Comment by laloyalo** [comment]
Hello @sylus , sorry to bother you with this old issue, but I just updated to the operator v0.1.0. Previous to that, I already had three namespaces with different ArgoCD CRDs installed, detected and running, but my way to do that with v0.0.15, was to modify the catalog source to watch for namespaces set to the empty string `""`, whenever I do that now, it throws an error mentioning that it cannot watch multiple namespaces.
```
time="2021-12-10T01:01:54Z" level=warning msg="issue getting operatorgroup target namespaces" error="TargetNamespaces cannot contain NamespaceAll: []" namespace=argocd operatorGroup=argocd-operator
```
I see that there have been several PRs to account to watch for multiple namespaces already, but I cannot get this to work.
Only the ArgoCD in the `argocd` namespace was updated to the latest release, the other namespaces are no longer being recognized by the ArgoCD operator to update them. 

What do you mean by:
> then was able to specify multiple namespaces

Where should I specify multiple namespaces?

One weird thing is that my ArgoCD operator installation is actually recognizing all of the other namespace instances, but for some reason it simply does not update them.

Does it have to do with the env var mentioned:

```yaml
- name: WATCH_NAMESPACE
    valueFrom:
      fieldRef:
        apiVersion: v1
        fieldPath: metadata.annotations['olm.targetNamespaces']
```

And the annotation set to:
```yaml
olm.targetNamespaces: argocd
```
Thanks in advance.

**Comment by laloyalo** [comment]
Sorry, I'll answer myself. I was able to fix it by editing the OperatorGroup and removing the spec, as described [here](https://docs.openshift.com/container-platform/4.3/operators/understanding_olm/olm-understanding-operatorgroups.html#olm-operatorgroups-target-namespace_olm-understanding-operatorgroups).

> Alternatively, you can omit both spec.selector and spec.targetNamespaces to specify a global OperatorGroup, which selects all namespaces:

```
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: argocd-operator
  namespace: argocd
```

Not sure if that's the way it's meant to be, but it worked and generated a Cluster Service Version (CSV) for all the namespaces, then all my other instances were updated as well.
